File: javascript/cote/PGMRS/큰_수_만들기.mjs
import {readFile} from 'fs/promises'

const TEST_CASES = 10
const MAX_NUMBER_LENGTH = 10000000

function solve(number, k) {
	const stack = []

	for (let c of number){
		while (k > 0 && !(stack.length == 0)) {
			if (parseInt(stack[stack.length - 1]) >= parseInt(c)) {
				break
			}
			stack.pop()
			k--
		}
		stack.push(c)
	}

	while (k > 0) {
		stack.pop()
		k--
	}

	return stack.join("")
}

const data = await readFile("data/큰_수_만들기.txt", {encoding : "utf-8"})
	.then(lines => lines.split("\n").map(line => line.trim()))

const testCases = data[0]
let cur = 1

for(let it = 0; it < testCases; it++) {
	const number = data[cur++]	
	const k = parseInt(data[cur++])
	const answer = parseInt(data[cur++])

	console.log(solve(number, k), answer)
}

for (let it = 0; it < TEST_CASES; it++) {
	const n = Math.floor(Math.random() * MAX_NUMBER_LENGTH - 1) + 1
	const number = new Array(n).fill(0)
		.map(i => Math.floor(Math.random() * 10))
		.join("")

	const k = Math.floor(Math.random() * n)

	console.log(solve(number, k))
}

File: java/src/main/java/com/example/cote/PGMRS/큰_수_만들기.java
package com.example.cote.PGMRS;

import static java.util.stream.Collectors.*;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;

public class 큰_수_만들기 {

  private static final int MAX_NUMBER_LENGTH = 1000000;

  private static final int TEST_CASES = 10;
  private static final Random random = new Random(43);

  private static String solve(String number, int k) {
    Stack<Character> stack = new Stack<>();

    for (char c : number.toCharArray()) {
      while (k > 0 && !stack.isEmpty() && c > stack.peek()) {
        stack.pop();
        k--;
      }
      stack.push(c);
    }

    while (k-- > 0) {
      stack.pop();
    }

    return stack.stream().map(String::valueOf).collect(joining());
  }

  public static void main(String... args) throws IOException {

    Path p = Paths.get(System.getProperty("user.dir") + "/data/큰_수_만들기.txt");
    BufferedReader reader = Files.newBufferedReader(p);

    int testCases = Integer.parseInt(reader.readLine());

    while (testCases > 0) {
      String number = reader.readLine();
      int k = Integer.parseInt(reader.readLine());
      String answer = reader.readLine();

      String ret = solve(number, k);

      if (!ret.equals(answer)) {
        System.out.printf("%s %s \n", answer, ret);
        return;
      }

      testCases--;
    }

    for (int it = 0; it < TEST_CASES; it++) {
      int n = random.nextInt(MAX_NUMBER_LENGTH - 2) + 2;

      String number =
          IntStream.range(0, n)
              .mapToObj(i -> Integer.toString(random.nextInt(10)))
              .collect(joining(""));

      int k = random.nextInt(n);

      System.out.println(solve(number, k));
    }
  }
}
